HW2/P5/P5.txt

```diff
@@ -0,0 +1,17 @@
+Multithreading with prange()
+With 1 thread we get results between 3.0 and 4.0 frames per second. With 4 threads we get results between 4.2 and 6.1 frames per second, which is consistently an improvement but almost double the range size. This is less than a 2x speedup for a 4x increase in threads.
+If prange() divides the loop into contiguous sections, then some sections will take longer than others because sub_update() checks all values from the current to the end and some contiguous sections will be closer to the end than others. Since sections take varying amounts of time, speedup is less than it could be. This also explains the runtime variability.
+
+Grid (spatial decomposition)
+With 1 thread we generally get results between 400 and 600 frames per second, with one result at 255. With 4 threads we get between 25 and 630 frames per second, with high variability.
+Adding the grid is ~100x faster than multithreading alone. For n=10,000, switching from an O(n^2) algorithm to an O(n) algorithm is likely to make a significant difference even if the O(n) algorithm has high overhead, e.g. updating the grid.
+The fact that 1 thread is generally ~5x faster than 4 threads makes no sense. After speaking with the course staff and attempting to debug, no error was found in the code and they agreed it made no sense. *Shrug* multithreading is complicated and perhaps my VM is not the best with threads + complex code.
```

Comment:
I had a similar experience where adding more threads actually showed a slowdown in the code. I also use a VM so that may be it. I did see spikes of numbers for 4 threads that were about 2x improvements over the 1 thread, but it would happen once every 4 print-outs so averaging with the smaller values would bring it back down.